core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala

```diff
@@ -0,0 +1,116 @@
+    val markersQueue = brokerStateMap(brokerId).markersQueue
+    markersQueue.add(txnMarkerRequest)
+    trace(s"Added markers $txnMarkerRequest for broker $brokerId")
+  }
+
+  def addRequestToSend(metadataPartition: Int, pid: Long, epoch: Short, result: TransactionResult, coordinatorEpoch: Int, topicPartitions: immutable.Set[TopicPartition]): Unit = {
+    val partitionsByDestination: immutable.Map[Int, immutable.Set[TopicPartition]] = topicPartitions.groupBy { topicPartition: TopicPartition =>
+      val leaderForPartition = metadataCache.getPartitionInfo(topicPartition.topic, topicPartition.partition)
+      val currentBrokers = mutable.Set.empty[Int]
+      leaderForPartition match {
+        case Some(partitionInfo) =>
+          val brokerId = partitionInfo.leaderIsrAndControllerEpoch.leaderAndIsr.leader
+          if (currentBrokers.add(brokerId)) {
+            // TODO: What should we do if we get BrokerEndPointNotAvailableException?
+            val broker = metadataCache.getAliveEndpoint(brokerId, interBrokerListenerName).get
+            addOrUpdateBroker(broker)
+          }
+          brokerId
+        case None =>
+          // TODO: there is a rare case that the producer gets the partition info from another broker who has the newer information of the
+          // partition, while the TC itself has not received the propagated metadata update; do we need to block when this happens?
+          throw new IllegalStateException("Cannot find the leader broker for the txn marker to write for topic partition " + topicPartition)
+      }
+    }
+
+    for ((brokerId: Int, topicPartitions: immutable.Set[TopicPartition]) <- partitionsByDestination) {
+      val txnMarker = new TxnMarkerEntry(pid, epoch, result, topicPartitions.toList.asJava)
+      addRequestForBroker(brokerId, CoordinatorEpochAndMarkers(metadataPartition, coordinatorEpoch, Utils.mkList(txnMarker)))
+    }
+  }
+
+  def maybeAddPendingRequest(metadataPartition: Int, metadata: TransactionMetadata): Boolean = {
+    val existingMetadataToWrite = pendingTxnMap.putIfAbsent(PendingTxnKey(metadataPartition, metadata.pid), metadata)
+    existingMetadataToWrite.isEmpty
+  }
+
+  def removeCompletedTxn(metadataPartition: Int, pid: Long): Unit = {
+    pendingTxnMap.remove(PendingTxnKey(metadataPartition, pid))
+  }
+
+  def pendingTxnMetadata(metadataPartition: Int, pid: Long): Option[TransactionMetadata] = {
+    pendingTxnMap.get(PendingTxnKey(metadataPartition, pid))
+  }
+
+  def clear(): Unit = {
+    brokerStateMap.clear()
+    pendingTxnMap.clear()
+  }
+
+  def removeStateForPartition(partition: Int): Unit = {
+    brokerStateMap.foreach {case(_, destinationAndQueue: DestinationBrokerAndQueuedMarkers) =>
+      val allMarkers: java.util.List[CoordinatorEpochAndMarkers] = new util.ArrayList[CoordinatorEpochAndMarkers] ()
+      destinationAndQueue.markersQueue.drainTo(allMarkers)
```

Comment:
They will error out with `NOT_COORDINATOR` on completion. Will that suffice? I'm not sure how we can cancel inflight requests in the `NetworkClient`